### ISSUE TITLE
Have unnecessary_overrides lint ignore overrides which include implementation comments

### ISSUE THREAD

**Issue by daiviknema** [dart-lang/linter#3512]
**Describe the issue**
It would be nice if the unnecessary_overrides rule could ignore cases such as the following:

```dart
@override
void doSomething() {
  super.doSomething();

  // TODO: add some more logic here.
}
```

```dart
@override
void doSomethingElse(MyUsefulArg arg) {
  super.doSomethingElse(arg);

  // No-op because of xyz reason.
}
```

```dart
// Please do not remove this override. It is needed it for xyz reason.
@override
void myAmazingMethod() {
  super.myAmazingMethod();
}
```

**Expected behavior**

If an implementation comment is present either on top of (or inside) an override that only calls the super method, then the unneccessary_override lint should ignore the override.

**Additional context**

This lint rule already ignores overrides that have documentation comments or annotations (https://dart-lang.github.io/linter/lints/unnecessary_overrides.html)

**Comment by srawlins** [comment]
I think these are all valid reasons to keep an override which appears to be unnecessary.

In the TODO case, I think this is better than a floating `// TODO` comment somewhere inside a class. I don't think we want to _parse_ comments anywhere ("Does it say TODO? Does it say FIXME?"). So I think the presence of a comment inside the body is enough reason to allow the unnecessary override.

As for the comment outside the body... that feels like we might be getter too lax. We could be reading a comment which is sort of... above the body but not really attached to it.

The risk of allowing these overrides, and not marking them unnecessary, is that we may be allowing them in unnecessarily. Someone may still be mis-understanding the override.

Another argument _against_ allowing these is that we already have a comment-based "ignore" system. Perhaps we should still required that `// ignore` comment. Allowing any textual comment may lead to other cases like:

```
// This is needed for memoization purposes.
bool? unusedField;
```

A user might ask, "can you not report that field as unused since I have a comment above it which explains its presence?"

I don't have a strong opinion either way, really.

**Comment by srawlins** [comment]
All that being said, I've found a few cases of the linter reporting a lint or not based on the presence of comments. Grep for precedingComments in `lib/`.

**Comment by bwilkerson** [comment]
For both of these suggestions, I think the question has to be whether more users would be harmed by the current behavior (of reporting code that currently does not have any semantic value) or whether more would be harmed by failing to report such cases.

Given that we have ignore comments to allow both of these use cases to be silenced, there is support today that allows users to not be harmed by the current behavior. The proposed behavior, on the other hand, has no work-around; user's can't signal that they want to be told anyway. So from a purely expressibility perspective, the current behavior appears to be better.

And, I'll point out that the current behavior is really not significantly different from the requested behavior. In both cases users are using a comment to indicate that the override is desired despite the rule. The only difference is the form and/or location of the comment.

I just can't see this change as an improvement.

**Comment by natebosch** [comment]
How about:

```diff
- @override
- void doSomething() {
-   super.doSomething();
- 
-   // TODO: add some more logic here.
- }
+ // TODO: override doSomething and add more logic.
```

```dart
// ignore: unnecessary_override; foo bar framework (deprecated) requires this
@override
void myAmazingMethod() {
  super.myAmazingMethod();
}
```

For the case with the method and an extra comment describing it as a "no op" - I would dig deeper into what pattern lead to the author thinking it would be necessary to document that.

I'm primarily concerned about cases which we don't consider technical debt where we would expect to add more instances of this pattern on an ongoing basis. Do we have any?

**Comment by daiviknema** [comment]
Thanks Nate and Brian, for the inputs. Nate's suggestions are good alternatives for what I proposed initially (and was how I handled the cleanup across google3). Closing this issue out!
